File: packages/sdk-flutter/lib/src/onramp.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

import 'storage.dart';

String _uuid() {
  final r = Random.secure();
  final bytes = List<int>.generate(16, (_) => r.nextInt(256));
  bytes[6] = (bytes[6] & 0x0f) | 0x40;
  bytes[8] = (bytes[8] & 0x3f) | 0x80;
  String hex(int b) => b.toRadixString(16).padLeft(2, '0');
  final h = bytes.map(hex).join();
  return '${h.substring(0, 8)}-${h.substring(8, 12)}-'
      '${h.substring(12, 16)}-${h.substring(16, 20)}-${h.substring(20)}';
}

final class OnRamp {
  OnRamp._();

  static const _sessionTimeoutMs = 30 * 60 * 1000;

  static String _apiKey = '';
  static String _host = '';
  static String? _appVersion;
  static String _anonymousId = '';
  static String _sessionId = '';
  static int _stepIndex = 0;
  static int _lastActiveMs = 0;

  // Tracks in-flight HTTP futures so flush() can await them.
  static final _inflight = <Future<void>>[];

  /// Overridden in tests to intercept HTTP calls without touching the network.
  @visibleForTesting
  static http.Client? httpClient;

  static Future<void> initialize({
    required String apiKey,
    required String host,
    String? appVersion,
  }) async {
    _apiKey = apiKey;
    _host = host.endsWith('/') ? host.substring(0, host.length - 1) : host;
    _appVersion = appVersion;

    final storedAnonId = await loadAnonymousId();
    if (storedAnonId != null) {
      _anonymousId = storedAnonId;
    } else {
      _anonymousId = _uuid();
      await saveAnonymousId(_anonymousId);
    }

    final stored = await loadSession();
    final now = DateTime.now().millisecondsSinceEpoch;
    if (stored != null && now - stored.lastActiveMs < _sessionTimeoutMs) {
      _sessionId = stored.id;
      _stepIndex = stored.stepCount;
      _lastActiveMs = stored.lastActiveMs;
    } else {
      _rotateSession();
    }
  }

  static void _rotateSession() {
    _sessionId = _uuid();
    _stepIndex = 0;
    _lastActiveMs = DateTime.now().millisecondsSinceEpoch;
    unawaited(saveSession(StoredSession(
      id: _sessionId,
      lastActiveMs: _lastActiveMs,
      stepCount: _stepIndex,
    )));
  }

  static void _refreshSession() {
    final now = DateTime.now().millisecondsSinceEpoch;
    if (_sessionId.isEmpty || now - _lastActiveMs > _sessionTimeoutMs) {
      _rotateSession();
    }
    _lastActiveMs = now;
  }

  static void step(String name, {Map<String, Object>? properties}) {
    if (_apiKey.isEmpty) {
      assert(false, '[OnRamp] call initialize() before tracking steps');
      return;
    }
    _refreshSession();
    final index = _stepIndex++;
    _lastActiveMs = DateTime.now().millisecondsSinceEpoch;
    unawaited(saveSession(StoredSession(
      id: _sessionId,
      lastActiveMs: _lastActiveMs,
      stepCount: _stepIndex,
    )));

    final event = <String, Object>{
      'schema_version': '1.0',
      'event_id': _uuid(),
      'event_type': 'step_entered',
      'app_key': _apiKey,
      'session_id': _sessionId,
      'anonymous_id': _anonymousId,
      'step_name': name,
      'step_index': index,
      'client_timestamp_ms': DateTime.now().millisecondsSinceEpoch,
      'platform': _platform,
      'os_version': _osVersion,
      'device_type': _deviceType,
    };
    if (_appVersion != null) event['app_version'] = _appVersion!;
    if (properties != null) event['properties'] = properties;
    _send(event);
  }

  /// Associate the current user with known traits (email, user ID, etc.).
  /// Call once after sign-in so connected integrations can match this user.
  ///
  /// ```dart
  /// OnRamp.identify({'email': user.email, 'userId': user.id});
  /// ```
  static void identify(Map<String, Object> traits) {
    if (_apiKey.isEmpty) {
      assert(false, '[OnRamp] call initialize() before calling identify()');
      return;
    }
    _refreshSession();

    final event = <String, Object>{
      'schema_version': '1.0',
      'event_id': _uuid(),
      'event_type': 'identify',
      'app_key': _apiKey,
      'session_id': _sessionId,
      'anonymous_id': _anonymousId,
      'step_name': '_identify',
      'step_index': 0,
      'client_timestamp_ms': DateTime.now().millisecondsSinceEpoch,
      'platform': _platform,
      'properties': traits,
    };
    if (_appVersion != null) event['app_version'] = _appVersion!;
    _send(event);
  }

  /// Force a new session. Call after sign-out so the next user starts fresh.
  static void newSession() => _rotateSession();

  /// Wait for all in-flight events to finish sending.
  /// Useful on app pause / before process exit.
  static Future<void> flush() async {
    if (_inflight.isEmpty) return;
    await Future.wait(List.from(_inflight));
    _inflight.clear();
  }

  /// Returns the current anonymous and session IDs so your server can
  /// associate backend events (purchases, trial starts) with this journey.
  static ({String anonymousId, String sessionId}) getIds() =>
      (anonymousId: _anonymousId, sessionId: _sessionId);

  static bool get isInitialized => _apiKey.isNotEmpty;

  @visibleForTesting
  static void reset() {
    _apiKey = '';
    _host = '';
    _appVersion = null;
    _anonymousId = '';
    _sessionId = '';
    _stepIndex = 0;
    _lastActiveMs = 0;
    _inflight.clear();
    httpClient = null;
  }

  static void _send(Map<String, Object> event) {
    final client = httpClient ?? http.Client();
    final f = client
        .post(
          Uri.parse('$_host/v1/events'),
          headers: {
            'Content-Type': 'application/json',
            'x-onramp-key': _apiKey,
          },
          body: jsonEncode({'events': [event]}),
        )
        .then((_) {})
        .catchError((_) {});
    _inflight.add(f);
    f.whenComplete(() => _inflight.remove(f));
  }

  static String get _platform {
    if (Platform.isIOS) return 'ios';
    if (Platform.isAndroid) return 'android';
    if (Platform.isMacOS) return 'macos';
    return 'other';
  }

  static String get _osVersion => Platform.operatingSystemVersion;

  static String get _deviceType {
    if (Platform.isMacOS || Platform.isLinux || Platform.isWindows) {
      return 'desktop';
    }
    return 'phone';
  }
}

File: packages/sdk-react-native/src/OnRamp.ts
import { OnRampClient } from '@onramp-sdk/core'
import { getOrCreateAnonymousId, loadSession, saveSession } from './storage.js'
import { AppState } from 'react-native'
import type { AppStateStatus } from 'react-native'
import { Platform } from 'react-native'

// Simple UUID v4 that works without crypto.randomUUID in older RN
function uuid(): string {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16)
  })
}

const DEFAULT_SESSION_TIMEOUT_MS = 30 * 60 * 1000           // 30 min
const DEFAULT_ANON_MAX_AGE_MS    = 365 * 24 * 60 * 60 * 1000 // 365 days

interface InitConfig {
  apiKey: string
  /** Your app's version string, e.g. "2.4.1". Shown in the dashboard breakdown. */
  appVersion?: string
  host?: string
  /** Quit-and-return within this window continues the same session. Default 30 min. */
  sessionTimeoutMs?: number
  /** Anonymous ID is rotated after this many milliseconds. Default 365 days. */
  anonymousIdMaxAgeMs?: number
}

let client: OnRampClient | null = null
let currentSessionId: string | null = null
let anonymousId: string | null = null
let stepCounter = 0
let sessionTimeoutMs = DEFAULT_SESSION_TIMEOUT_MS
let lastActive = 0

// The OS the app actually runs on. This is the user's *device*, distinct from
// the SDK framework (React Native) which we report separately as `_framework`.
function getPlatform(): 'ios' | 'android' | 'web' | 'other' {
  if (Platform.OS === 'ios') return 'ios'
  if (Platform.OS === 'android') return 'android'
  if (Platform.OS === 'web') return 'web'
  return 'other'
}

// Just the OS version, e.g. "17.2" (iOS) or "34" (Android API level) - the OS
// name lives in `platform`, so we no longer prefix it here.
function getOsVersion(): string {
  return String(Platform.Version)
}

function getDeviceType(): 'phone' | 'tablet' | 'desktop' {
  if (Platform.isTV) return 'desktop'
  if (Platform.OS === 'ios' && Platform.isPad) return 'tablet'
  return 'phone'
}

function persistSession(): void {
  if (!currentSessionId) return
  saveSession({ id: currentSessionId, lastActive, stepCounter })
}

function rotateSession(): void {
  currentSessionId = uuid()
  stepCounter = 0
  lastActive = Date.now()
  persistSession()
}

/** Rotate if the user has been away longer than the session timeout. */
function ensureSessionFresh(): void {
  if (!currentSessionId || Date.now() - lastActive > sessionTimeoutMs) {
    rotateSession()
  }
}

export const OnRamp = {
  async init(config: InitConfig): Promise<void> {
    sessionTimeoutMs = config.sessionTimeoutMs ?? DEFAULT_SESSION_TIMEOUT_MS
    anonymousId = await getOrCreateAnonymousId(uuid, config.anonymousIdMaxAgeMs ?? DEFAULT_ANON_MAX_AGE_MS)

    // Resume the previous session if the user came back within the timeout
    const stored = await loadSession()
    if (stored && Date.now() - stored.lastActive < sessionTimeoutMs) {
      currentSessionId = stored.id
      stepCounter = stored.stepCounter
      lastActive = stored.lastActive
    } else {
      rotateSession()
    }

    client = new OnRampClient({
      apiKey: config.apiKey,
      host: config.host,
      platform: getPlatform(),
      framework: 'react_native',
      appVersion: config.appVersion ?? null,
      uuidFn: uuid,
    })

    AppState.addEventListener('change', (state: AppStateStatus) => {
      if (state === 'background' || state === 'inactive') {
        lastActive = Date.now()
        persistSession()
        client?.flush()
      } else if (state === 'active') {
        ensureSessionFresh()
      }
    })
  },

  step(
    stepName: string,
    options?: {
      /** Custom event properties (e.g. { club_count: 5, user_type: 'free' }) */
      properties?: Record<string, string | number | boolean>
      /** @internal Used by NavigationTracker - not part of the public API */
      _eventType?: 'nav_entered'
    }
  ): void {
    if (!client || !currentSessionId || !anonymousId) {
      if (__DEV__) console.warn('[OnRamp] call OnRamp.init() before tracking steps')
      return
    }

    ensureSessionFresh()
    const index = stepCounter++
    lastActive = Date.now()
    persistSession()

    client.track({
      sessionId: currentSessionId,
      anonymousId,
      eventType: options?._eventType,
      stepName,
      stepIndex: index,
      osVersion: getOsVersion(),
      deviceType: getDeviceType(),
      properties: options?.properties ?? null,
    })
  },

  /**
   * Associate the current user with known traits (email, user ID, etc.).
   * Call once after sign-in so connected integrations (Stripe, RevenueCat)
   * can match this user to their records.
   *
   * ```ts
   * OnRamp.identify({ email: user.email, userId: user.id })
   * ```
   */
  identify(traits: Record<string, string | number | boolean>): void {
    if (!client || !currentSessionId || !anonymousId) {
      if (__DEV__) console.warn('[OnRamp] call OnRamp.init() before calling identify()')
      return
    }
    client.identify({ sessionId: currentSessionId, anonymousId, traits })
  },

  newSession(): void {
    rotateSession()
  },

  flush(): Promise<void> {
    return client?.flush() ?? Promise.resolve()
  },

  /**
   * Return the current anonymous and session IDs so your server can associate
   * backend events (purchases, trial starts, etc.) with this user's journey.
   * Pass both in the request body at the moment of the action — session IDs rotate.
   */
  getIds(): { anonymousId: string | null; sessionId: string | null } {
    return { anonymousId, sessionId: currentSessionId }
  },

  get isInitialized(): boolean {
    return client !== null
  },
}
